File: widgetssdk/src/main/java/com/glia/widgets/call/Configuration.java
package com.glia.widgets.call;

import com.glia.androidsdk.Engagement;
import com.glia.widgets.core.configuration.GliaSdkConfiguration;
import com.glia.widgets.helper.Utils;

public class Configuration {
    private final GliaSdkConfiguration sdkConfiguration;
    private final Engagement.MediaType mediaType;
    private final boolean isUpgradeToCall;

    private Configuration(Builder builder) {
        this.sdkConfiguration = builder.widgetsConfiguration;
        this.mediaType = builder.mediaType;
        this.isUpgradeToCall = builder.isUpgradeToCall != null ? builder.isUpgradeToCall : false;
    }

    public GliaSdkConfiguration getSdkConfiguration() {
        return this.sdkConfiguration;
    }

    public Engagement.MediaType getMediaType() {
        return this.mediaType;
    }

    public boolean getIsUpgradeToCall() {
        return this.isUpgradeToCall;
    }

    public static class Builder {
        private GliaSdkConfiguration widgetsConfiguration;
        private Engagement.MediaType mediaType;
        private Boolean isUpgradeToCall;

        public Builder setWidgetsConfiguration(GliaSdkConfiguration configuration) {
            this.widgetsConfiguration = configuration;
            return this;
        }

        public Builder setMediaType(Engagement.MediaType mediaType) {
            this.mediaType = mediaType;
            return this;
        }

        public Builder setMediaType(String mediaType) {
            this.mediaType = Utils.toMediaType(mediaType);
            return this;
        }

        public Builder setIsUpgradeToCall(Boolean isUpgradeToCall) {
            this.isUpgradeToCall = isUpgradeToCall;
            return this;
        }

        public Builder() {
            this.isUpgradeToCall = false;
            this.mediaType = Engagement.MediaType.AUDIO;
        }

        public Configuration build() {
            return new Configuration(this);
        }

        public static Builder from(Configuration configuration) {
            return builder()
                    .setWidgetsConfiguration(configuration.sdkConfiguration)
                    .setMediaType(configuration.mediaType)
                    .setIsUpgradeToCall(configuration.isUpgradeToCall);
        }

        public static Builder builder() {
            return new Builder();
        }
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/call/Configuration.kt
package com.glia.widgets.call

import com.glia.androidsdk.Engagement
import com.glia.widgets.core.configuration.GliaSdkConfiguration
import com.glia.widgets.helper.Utils

internal class Configuration private constructor(builder: Builder) {
    @JvmField
    val sdkConfiguration: GliaSdkConfiguration?
    @JvmField
    val mediaType: Engagement.MediaType?
    @JvmField
    val isUpgradeToCall: Boolean

    init {
        sdkConfiguration = builder.widgetsConfiguration
        mediaType = builder.mediaType
        isUpgradeToCall = builder.isUpgradeToCall
    }

    class Builder {
        var widgetsConfiguration: GliaSdkConfiguration? = null
            private set
        var mediaType: Engagement.MediaType? = null
            private set
        var isUpgradeToCall = false
            private set

        fun setWidgetsConfiguration(configuration: GliaSdkConfiguration?) = apply {
            widgetsConfiguration = configuration
        }

        fun setMediaType(mediaType: Engagement.MediaType?) = apply {
            this.mediaType = mediaType
        }

        fun setMediaType(mediaType: String) = apply {
            this.mediaType = Utils.toMediaType(mediaType)
        }

        fun setIsUpgradeToCall(isUpgradeToCall: Boolean) = apply {
            this.isUpgradeToCall = isUpgradeToCall
        }


        fun build(): Configuration {
            return Configuration(this)
        }
    }
}
